actors/builtin/market/testing.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	var pendingProposal DealProposal
+	err = pendingProposals.ForEach(&pendingProposal, func(key string) error {
+		proposalCID, err := cid.Parse([]byte(key))
+		if err != nil {
+			return err
+		}
+
+		pcid, err := pendingProposal.Cid()
+		if err != nil {
+			return err
+		}
+		acc.Require(pcid.Equals(proposalCID), "pending proposal's key does not match its CID %v != %v", pcid, proposalCID)
+
+		_, found := proposalCids[pcid]
+		acc.Require(found, "pending proposal with cid %v not fond within proposals %v", pcid, pendingProposals)
+
+		pendingProposalCount++
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	//
+	// Escrow Table and Locked Table
+	//
+
+	escrowTable, err := adt.AsBalanceTable(store, st.EscrowTable)
+	if err != nil {
+		return nil, acc, err
+	}
+
+	lockTableCount := uint64(0)
+	lockTable, err := adt.AsBalanceTable(store, st.LockedTable)
+	if err != nil {
+		return nil, acc, err
+	}
+	var lockedAmount abi.TokenAmount
+	lockedTotal := abi.NewTokenAmount(0)
+	err = (*adt.Map)(lockTable).ForEach(&lockedAmount, func(key string) error {
+		addr, err := address.NewFromBytes([]byte(key))
+		if err != nil {
+			return err
+		}
+		lockedTotal = big.Add(lockedTotal, lockedAmount)
+
+		// every entry in locked table should have a corresponding entry in escrow table that is at least as high
+		escrowAmount, err := escrowTable.Get(addr)
+		if err != nil {
+			return err
+		}
+		acc.Require(escrowAmount.GreaterThanEqual(lockedAmount),
+			"locked funds for %s, %s, greater than escrow amount, %s", addr, lockedAmount, escrowAmount)
+
+		lockTableCount++
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	// lockTable total should be sum of client and provider locked plus client storage fee
+	expectedLockTotal := big.Sum(st.TotalProviderLockedCollateral, st.TotalClientLockedCollateral, st.TotalClientStorageFee)
+	acc.Require(lockedTotal.Equals(expectedLockTotal),
+		"locked total, %s, does not sum to provider locked, %s, client locked, %s, and client storage fee, %s",
+		lockedTotal, st.TotalProviderLockedCollateral, st.TotalClientLockedCollateral, st.TotalClientStorageFee)
+
+	//
+	// Deal Ops by Epoch
+	//
+
+	dealOpEpochCount := uint64(0)
+	dealOpCount := uint64(0)
+	dealOps, err := AsSetMultimap(store, st.DealOpsByEpoch)
+	if err != nil {
+		return nil, acc, err
+	}
+
+	// get into internals just to iterate through full data structure
+	var setRoot cbg.CborCid
+	err = dealOps.mp.ForEach(&setRoot, func(key string) error {
+		epoch, err := binary.ReadUvarint(bytes.NewReader([]byte(key)))
+		if err != nil {
+			return errors.Wrapf(err, "deal ops has key that is not an int: %s", key)
+		}
+
+		dealOpEpochCount++
+		return dealOps.ForEach(abi.ChainEpoch(epoch), func(id abi.DealID) error {
+			_, found := allIDs[id]
+			acc.Require(found, "deal op found for deal id %d with missing proposal at epoch %d", id, epoch)
+			dealOpCount++
+			return nil
```

Comment:
Either there should exist a corresponding deal state, or `CurrEpoch() >= deal.StartEpoch`. We can accept a current epoch parameter here too.